src/README ImgPreprocess.md

```diff
@@ -0,0 +1,240 @@
+                elif (n[r - 1, c - 1] == -2 and self.nim[r, c] == 0):
+                    # Convert to black if the white pixel is surrounded by 2 
+                    # black pixels; one in the same column and other in the same row
+                    # as the white pixel
+                    if self.nim[r - 1, c] + self.nim[r, c - 1] == -2 or self.nim[r + 1, c] + self.nim[r, c - 1] == -2 or self.nim[r + 1, c] + self.nim[r, c + 1] == -2 or self.nim[r - 1, c] + self.nim[r, c + 1] == -2: self.nim[r, c] = -1
+
+    if way == 0: 
+        sharpen_round_edge()
+        sharpen_corners()
+    elif way == 1:
+        sharpen_round_edge()
+    elif way == 2:
+        sharpen_corners()
+    self.reduceMatrix(margin)
+```
+
+#### 7. The extra space present on the borders of the image matrix are trimmed by passing the no. of pixels to be deleted.
+
+## [`trim()`](https://github.com/tushdon2/RL-based-Maze-Solver/blob/fbd11ce17a93b9e0d79abbe9c3acc61b46c69b3a/src/ImgPreprocess.py#L137)
+_Trims the maze by **pix** pixels (default: 1 px) from all sides._
+```
+def trim(self, pix = 1):
+    r = []
+    c = []
+    for i in range(pix):
+        r.append(i)
+        c.append(i)
+    for i in range(pix, 0, -1):
+        r.append(self.h - i)
+        c.append(self.w - i)
+
+    self.nim = np.delete(np.delete(self.nim, r, 0), c, 1)
+    self.h -= 2 * pix
+    self.w -= 2 * pix
+```
+
+#### 8. The openings in the border of the maze are detected, and one of them is defined as starting and other as the ending.
+## [`detectOpenings()`](https://github.com/tushdon2/RL-based-Maze-Solver/blob/fbd11ce17a93b9e0d79abbe9c3acc61b46c69b3a/src/ImgPreprocess.py#L153)
+_Detects the openings (start and the end points) in the maze. Ought to trim outer whitespaces and reduce the matrix using `trim()` and `reduceMatrix()` operation respectively before using this._
+```
+def detectOpenings(self):
+    # Detect opening in first in last row
+    for i in range(self.w):
+        if self.nim[0, i] == 0: 
+            l.append(0)
+            l.append(i)
+            count += 1
+            break
+
+    for i in range(self.w):
+        if self.nim[self.h - 1, i] == 0: 
+            l.append(self.h - 1)
+            l.append(i)
+            count += 1
+            if count == 2 : return l        
+
+    # Detect opening in the first and last column
+    for i in range(self.h):
+        if self.nim[i, 0] == 0: 
+            l.append(i)
+            l.append(0)
+            count += 1
+            if count == 2 : return l
+    
+    for i in range(self.h):
+        if self.nim[i, self.w -1] == 0: 
+            l.append(i)
+            l.append(self.w -1)
+            count += 1
+            if count == 2 : return l
+```
+
+#### 9. This [ImgPreprocess.py](https://github.com/tushdon2/RL-based-Maze-Solver/blob/main/src/ImgPreprocess.py) file is then imported by the [Maze Solving Code(Code_4.0.ipynb)](https://github.com/tushdon2/RL-based-Maze-Solver/blob/main/src/Code_4.0.ipynb).
```

Comment:
Do necessary changes here, after renaming the files as suggested in the previous comment.

Reply:
@SanjeevKrishnan  
We have made all the changes you asked for.
Please review.